include/stream-client/logger.hpp

```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include <chrono>
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+namespace stream_client {
+
+/**
+ * Base enum of log log levels
+ */
+enum class log_level {
+    error = 0,
+    warning,
+    info,
+};
+
+/**
+ * Base class for log events from internals of library
+ */
+class log_interface {
+public:
+    virtual ~log_interface() {}
+
+    /**
+     * Message from library for application.
+     *
+     * @param[in] level Level of message, application may use correct it according the target if necessary.
+     * @param[in] target Target of message is used for filter messages from different parts of library.
+     * @param[in] message The body and main reason of message.
+     * @param[in] error The error instance if exists.
+     *
+     * @note It can be from multiple threads, you should worry about sync and not lock this calls.
+     */
+    virtual void message(log_level level, const std::string& target, const std::string& message, const std::runtime_error* error = nullptr) const = 0;
```

Comment:
`location` is a much better name for the `target` parameter.
`error` what for? If we are logging inside try/catch block we can just call `message(log_level::error, exc.what());` and such usage would be more forgiving. Exceptions are rare to be passed as an argument, also such signature forbids anything not directly inherited from `std::runtime_error` and assumes that `what()` method is available. It's not always a case, anything can be thrown.